### PR TITLE
add latex extension example with mathjax

### PR DESCRIPTION
**Pull request by IuryPiva**
- [x] issues: relates #6031 
- [ ] tests added / passed
- [x] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by tcmetzger** [comment]
I have renamed and moved some LaTeX-related example files (see https://github.com/bokeh/bokeh/pull/11023#issuecomment-797805495).
I'm planning to update the respective readme.md files in the folders for the LaTeX extension folder once we have support for LaTeX in Bokeh itself to point potential users towards those new capabilities.

**Comment by bryevdv** [comment]
LGTM

File: examples/models/file/latex_extension_mathjax.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  Float
+)
+from bokeh.plotting import figure, output_file, show
+from bokeh.util.compiler import TypeScript
+
+output_file('latex_extension_with_mathjax.html')
+
+class LatexLabel(Label):
+    width = Nullable(Float, help="""
+    The gradient of the line, in data units
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:17]: The docstring needs updating. The property shouldn't be nullable.

File: examples/models/file/latex_extension_mathjax.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    height = Nullable(Float, help="""
+    The gradient of the line, in data units
+    """)
+    """A subclass of `Label` with all of the same class attributes except
+    canvas mode isn't supported and DOM manipulation happens in the TypeScript
+    superclass implementation that requires setting `render_mode='css'`).
+
+    Only the render method of LabelView is overwritten to perform the
+    text -> latex (via MathJax) conversion
+    """
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:28]: Docstring should should come first in a class.

File: examples/models/file/latex_extension_mathjax.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    __javascript__ = ["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg.js"]
+    __implementation__ = TypeScript("""
+import {Label, LabelView} from "models/annotations/label"
+import * as p from "core/properties"
+
+declare namespace MathJax {
+  function tex2svg(input: string): any
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:35]: Use a specific return type. `any` is not acceptable in this codebase, unless all other options were exhausted.

File: examples/models/file/latex_extension_mathjax.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    const xscale = this.coordinates.x_scale
+    const yscale = this.coordinates.y_scale
+
+    let sx = this.model.x_units == "data" ? xscale.compute(this.model.x) : panel.bbox.xview.compute(this.model.x)
+    let sy = this.model.y_units == "data" ? yscale.compute(this.model.y) : panel.bbox.yview.compute(this.model.y)
+
+    sx += this.model.x_offset
+    sy -= this.model.y_offset
+
+    const svgElement = MathJax.tex2svg(this.model.text).children[0]
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:53]: We use snake case. Examples are no exception.

File: examples/models/file/latex_extension_mathjax.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    let outerHTML = svgElement.outerHTML,
+      blob = new Blob([outerHTML],{type:'image/svg+xml;charset=utf-8'});
+
+    let URL = window.URL || window.webkitURL || window;
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:58]: Is this really necessary? It seems that all browsers we support (including IE), support `URL.createObjectURL()`. Also, which browser/spec allows for access through `window`?

File: examples/models/file/latex_extension_mathjax.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    let blobURL = URL.createObjectURL(blob);
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:59]: Use `const` wherever applicable.

Comment (mattpap) [reply examples/models/file/latex_extension_mathjax.py:59]: Make semicolons consistent with bokehjs, by taking advantage of semicolon insertion (i.e. use semicolons when unavoidable).

File: examples/models/file/latex_extension_mathjax.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    let image = new Image();
+
+    image.onload = () => {
+      this.layer.ctx.drawImage(image, sx, sy, this.model.width, this.model.height);
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:64]: Painting just like this will interfere with z-ordering of renderers. You need to request a repaint when an image is loaded. Unless you can guarantee that this callback triggers immediately in all supported browsers (I'm guessing given that we're dealing with data URLs here). If that's not the case, then additionally bokehjs' needs to be told that this is a delayed render, so that it doesn't notify its idle state too early (search for `notify_finished` and related logic).

File: examples/models/file/latex_extension_mathjax.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+from bokeh.models import Label
+from bokeh.palettes import Spectral4
+from bokeh.core.properties import (
+  Nullable,
+  Float
+)
+from bokeh.plotting import figure, output_file, show
+from bokeh.util.compiler import TypeScript
+
+output_file('latex_extension_with_mathjax.html')
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:13]: Examples should not use `output_file()`, as it may interfere with examples' discovery logic.

File: examples/models/file/latex_extension_mathjax.py (hunk 4 above)
Comment (mattpap) [reply examples/models/file/latex_extension_mathjax.py:53]: I'm repeating myself (see `any`), but what exactly this function returns, a collection of SVG elements? Can it return an empty collection? Either entire spectrum of return values needs to be handled, or there should be an assertion and/or comment regarding how this works, so that we don't end up with a bogus exception in some cases.

File: examples/models/file/latex_extension_mathjax.py (hunk 5 above)
Comment (mattpap) [reply examples/models/file/latex_extension_mathjax.py:58]: Additionally, you shouldn't access globals through `window`, because it's redundant and this code may run (in future) in an environment where the global context is not `window` (e.g. a web worker global context). If you need to check if an object exists globally, like this `URL`, then use `typeof URL !== "undefined"`.

File: examples/models/file/latex_extension_mathjax.py (hunk 7 above)
Comment (IuryPiva) [reply examples/models/file/latex_extension_mathjax.py]: I see the possible issue, But I didn't find anything on how to tell Bokeh that something is still waiting to be finished. And if I request a repaint after the loading of the image it starts blinking. Could you give me an example or directions where to go from here?

Comment (IuryPiva) [reply examples/models/file/latex_extension_mathjax.py]: @mattpap

Comment (mattpap) [reply examples/models/file/latex_extension_mathjax.py]: `ImageURL` is the best example. Look for `request_render()` (`ImageLoader.loaded`), `has_finished()` and `notify_finished()` (`has_finished()` is particularly important).

File: examples/models/file/latex_extension_mathjax.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    let sx = this.model.x_units == "data" ? xscale.compute(this.model.x) : panel.bbox.xview.compute(this.model.x)
+    let sy = this.model.y_units == "data" ? yscale.compute(this.model.y) : panel.bbox.yview.compute(this.model.y)
+
+    sx += this.model.x_offset
+    sy -= this.model.y_offset
+
+    const svg_element = MathJax.tex2svg(this.model.text).children[0] as SVGElement
+
+    const outer_HTML = svg_element.outerHTML,
+      blob = new Blob([outer_HTML],{type:'image/svg+xml;charset=utf-8'})
```
Comment (bryevdv) [inline examples/models/file/latex_extension_mathjax.py:55]:
```suggestion
      blob = new Blob([outer_HTML], {type:'image/svg+xml;charset=utf-8'})
```

Comment (bryevdv) [reply examples/models/file/latex_extension_mathjax.py:55]: indentation and needs `const` afaik?

File: examples/models/file/latex_extension_mathjax.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+p.x_range.range_padding = 0
+
+x = np.arange(0.0, 20.0, 0.02)
+
+for i, n in enumerate([0, 1, 4, 7]):
+    p.line(x, jv(n, x), line_width=3, color=Spectral4[i], alpha=0.8, legend_label="𝜈=%d" % n)
+
+text = ("x = {-b \pm \sqrt{b^2-4ac} \over 2a}")
+
+latex = LatexLabel(text=text,x=4.5, y=250, x_units='data', y_units='screen',
```
Comment (bryevdv) [inline examples/models/file/latex_extension_mathjax.py:108]:
```suggestion
latex = LatexLabel(text=text, x=4.5, y=250, x_units='data', y_units='screen',
```

**Comment by bryevdv** [comment]
I only had a couple of minor comments. Would be nice to get this merged in the next few days

File: examples/models/file/latex_extension_mathjax.py (hunk 9 above)
Comment (mattpap) [reply examples/models/file/latex_extension_mathjax.py:55]: This syntax is correct JS, but preferably there should be another `const`.

File: examples/models/file/latex_extension_mathjax.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+    let sx = this.model.x_units == "data" ? xscale.compute(this.model.x) : panel.bbox.xview.compute(this.model.x)
+    let sy = this.model.y_units == "data" ? yscale.compute(this.model.y) : panel.bbox.yview.compute(this.model.y)
+
+    sx += this.model.x_offset
+    sy -= this.model.y_offset
+
+    const svg_element = MathJax.tex2svg(this.model.text).children[0] as SVGElement
+    const outer_HTML = svg_element.outerHTML
+    const blob = new Blob([outer_HTML],{type:'image/svg+xml;charset=utf-8'})
```
Comment (bryevdv) [inline examples/models/file/latex_extension_mathjax.py:54]:
```suggestion
    const blob = new Blob([outer_HTML], {type:'image/svg+xml;charset=utf-8'})
```

nit: we always put space between function params

Comment (bryevdv) [reply examples/models/file/latex_extension_mathjax.py:54]: I'm going to go ahead and commit this via the UI to kick off another build (failures were Conda network issues hopefully should all go green now)

**Review by bryevdv — APPROVED** [review]

File: examples/models/file/latex_extension_mathjax.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    const blob = new Blob([outer_HTML], {type:'image/svg+xml;charset=utf-8'})
+    const blob_URL = URL.createObjectURL(blob)
+
+    const image = new Image()
+
+    image.onload = () => {
+      this.layer.ctx.drawImage(image, sx, sy, this.model.width, this.model.height)
+
+      this.notify_finished()
+    }
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:63]: URL object should be revoked when not needed anymore.

File: examples/models/file/latex_extension_mathjax.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+    const svg_element = MathJax.tex2svg(this.model.text).children[0] as SVGElement
+    const outer_HTML = svg_element.outerHTML
+    const blob = new Blob([outer_HTML], {type:'image/svg+xml;charset=utf-8'})
+    const blob_URL = URL.createObjectURL(blob)
+
+    const image = new Image()
+
+    image.onload = () => {
+      this.layer.ctx.drawImage(image, sx, sy, this.model.width, this.model.height)
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:60]: It would be better to infer `width` and `height` by default, and leave those properties for optional scaling.

File: examples/models/file/latex_extension_mathjax.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+      this.notify_finished()
```
Comment (mattpap) [inline examples/models/file/latex_extension_mathjax.py:62]: So where are we with deciding whether `onload`  triggers immediately in this context? Because if it does trigger immediately, then notification is not needed. If it doesn't, then `drawImage()` needs to be replaced with `request_paint()` with logic preventing infinite repaint loop.

Comment (mattpap) [reply examples/models/file/latex_extension_mathjax.py:62]: On its own, this `notify_finished()` does nothing, as notification was already sent. One has to change `has_finished` method to make sure that notification is not sent before this line. See `ImageURL.has_finished`. This is related to the above point about preventing infinite paint loop.

File: examples/models/file/latex_extension_mathjax.py (hunk 13 above)
Comment (bryevdv) [reply examples/models/file/latex_extension_mathjax.py:60]: I agree, but this is just an example at this point. I think it's OK as-is for now.

File: examples/models/file/latex_extension_mathjax.py (hunk 14 above)
Comment (bryevdv) [reply examples/models/file/latex_extension_mathjax.py:62]: > So where are we with deciding whether onload triggers immediately in this context? 

Is there a reference to a question/issue I can go look at?

Comment (mattpap) [reply examples/models/file/latex_extension_mathjax.py:62]: Earlier in this thread (https://github.com/bokeh/bokeh/pull/11067#discussion_r595576100).

Comment (bryevdv) [reply examples/models/file/latex_extension_mathjax.py:62]: Sorry the wording made it sound like you were asking about a policy decision inside our codebase. I am not sure there is any way to know for sure easily. The only reference I could find was from ~10 years ago, and seemed to indicate (from empirical testing) "usually but not always":  https://stackoverflow.com/a/4807114/3406693 But it's possible that things have changed in the interim. Still if it is not documented in a spec somewhere I think we have to play it safe and assume the answer is No

Comment (IuryPiva) [reply examples/models/file/latex_extension_mathjax.py:62]: > On its own, this notify_finished() does nothing, as notification was already sent. One has to change

working on this

**Comment by tcmetzger** [comment]
The rendered result looks great to me:
grafik

@bryevdv Do you know what the CI errors mean? They seem to be documentation-related - but I was able to locally build the docs off this branch without any issues.

**Comment by bryevdv** [comment]
@IuryPiva the same errors are showing up in other PRs it appears some CI dependency (maybe networkx?) has made a change that affects us. TLDR you don't need to worry about them. :)

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
@mattpap I believe the previous issues are all addressed, I will plan to merge this tomorrow unless there is comment

**Review by bryevdv — APPROVED** [review]
